astropy/utils/masked/core.py

```diff
@@ -41,7 +42,123 @@
 """.format
 
 
-class Masked(NDArrayShapeMethods):
+class MaskMethods:
+    """Mixin class providing class methods for dealing with Masked inputs.
+
+    Mostly useful as part of `~astropy.utils.masked.MaskableShapedLikeNDArray`,
+    but also used in `~astropy.utils.masked.Masked`.
+    """
+
+    @classmethod
+    def _get_data_and_mask(cls, data, allow_ma_masked=False):
+        """Split data into unmasked and mask, if present.
+
+        Parameters
+        ----------
+        data : array-like
+            Possibly masked item, judged by whether it has a ``mask`` attribute.
+            If so, checks for being an instance of `~astropy.utils.masked.Masked`
+            or `~numpy.ma.MaskedArray`, and gets unmasked data appropriately.
+        allow_ma_masked : bool, optional
+            Whether or not to process `~numpy.ma.masked`, i.e., an item that
+            implies no data but the presence of a mask.
+
+        Returns
+        -------
+        unmasked, mask : array-like
+            Unmasked will be `None` for `~numpy.ma.masked`.
+
+        Raises
+        ------
+        ValueError
+            If `~numpy.ma.masked` is passed in and ``allow_ma_masked`` is not set.
+
+        """
+        mask = getattr(data, "mask", None)
+        if mask is not None:
+            try:
+                data = data.unmasked
+            except AttributeError:
+                if not isinstance(data, np.ma.MaskedArray):
+                    raise
+                if data is np.ma.masked:
+                    if allow_ma_masked:
+                        data = None
+                    else:
+                        raise ValueError("cannot handle np.ma.masked here.") from None
+                else:
+                    data = data.data
+
+        return data, mask
+
+    @classmethod
+    def _get_data_and_masks(cls, *args):
```

Comment:
Maybe also use `arrays` instead of `args` if this will be public. Overall in some of these functions I feel like `array` is a better name than `data` since that name narrows the type considerably. But that may be sliding into "perfect is the enemy of good" territory.

Reply:
Sorry, should have answered here: I ended up thinking that the plural was not needed as much (or at least it wasn't clear it was going to be needed a lot), so just left this on the class to be used by `__array_ufunc__`.

Reply:
p.s. Did go with `array` in the general `get_data_and_mask`